### PR TITLE
[generate_opcheck_tests] Always print a repro

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109641
* __->__ #109640
* #109639
* #109638
* #109637

On failure of a test, we will always print a "repro". This repro isn't
really runnable but gives the user a sense of how to actually reproduce
the test without the test suite, because using the test suite is a bit
convoluted.

If the user passes PYTORCH_OPCHECK_PRINT_BETTER_REPRO, we will print a
fuller repro that saves the exact problematic test inputs to disk and
reads them back out.

Test Plan:
- expecttests on the generate_repro helper function
- tried this out locally.

### CODE REVIEW

File: torch/testing/_internal/optests/generate_tests.py [hunk 1]
```diff
@@ -445,16 +441,16 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
                     self.seen_ops_to_errors[qualname] = []
                 self.run_test_util(func, args_c, kwargs_c)
             except Exception as ex:
-                if should_print_repro():
+                if should_print_better_repro():
                     self.seen_ops_to_errors[qualname].append((ex, func, args, kwargs))
                 else:
-                    self.seen_ops_to_errors[qualname].append((ex, None, None, None))
+                    self.seen_ops_to_errors[qualname].append((ex, func, None, None))
```
Comment (bdhirsh) [inline torch/testing/_internal/optests/generate_tests.py:447]: nit: the comment for `seen_ops_to_errors` look out of date?
```
# Maps qualname -> List[exception]
```

**Review by bdhirsh — APPROVED** [review]
very cool

File: torch/testing/_internal/optests/generate_tests.py (hunk 1 above)
Comment (zou3519) [reply torch/testing/_internal/optests/generate_tests.py:447]: good catch, fixed in the PR one above this one

File: torch/testing/_internal/optests/generate_tests.py [hunk 2]
```diff
@@ -445,16 +441,16 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
...
         elif option == "skip":
             pass
         return result
 
 
-def should_print_repro() -> None:
+def should_print_better_repro() -> None:
```
Comment (soulitzer) [inline torch/testing/_internal/optests/generate_tests.py:453]: nit: Looks like the comment for this function still says 
```
"""If set, the tests generated by `generate_opcheck_tests` will print a
    repro command on failure.
```

**Review by soulitzer — APPROVED** [review]

**Pull request closed without merging**